File: src/sim_rtu/points/store.py
"""PointStore: central value registry for simulated BACnet points.

Provides thread-safe read/write access by either:
- volttron_name (string key)
- BACnet object key (bacnet_object_type, index) tuple

Writable points use a PriorityArray; read-only points store a single value.
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass, field

from sim_rtu.points.priority_array import PriorityArray
from sim_rtu.points.registry import PointDefinition


@dataclass
class _PointSlot:
    """Internal storage for a single point's value and metadata."""

    definition: PointDefinition
    priority_array: PriorityArray | None
    value: float | None


@dataclass
class PointStore:
    """Central value registry for all points in a simulated device.

    Initialize from a list of PointDefinition objects. Supports concurrent
    access via an asyncio.Lock.
    """

    _by_name: dict[str, _PointSlot] = field(default_factory=dict, init=False)
    _by_object_key: dict[tuple[str, int], _PointSlot] = field(
        default_factory=dict, init=False
    )
    _lock: asyncio.Lock = field(default_factory=asyncio.Lock, init=False)

    @classmethod
    def from_definitions(cls, definitions: list[PointDefinition]) -> PointStore:
        """Create a PointStore from a list of point definitions.

        Writable points get a PriorityArray initialized with the default value
        at priority 16. Read-only points store the default value directly.
        """
        store = cls()

        for defn in definitions:
            if defn.writable:
                pa = PriorityArray(default_value=defn.default_value)
                if defn.default_value is not None and defn.write_priority is not None:
                    pa.write(defn.write_priority, defn.default_value)
                slot = _PointSlot(
                    definition=defn,
                    priority_array=pa,
                    value=None,
                )
            else:
                slot = _PointSlot(
                    definition=defn,
                    priority_array=None,
                    value=defn.default_value,
                )

            store._by_name[defn.volttron_name] = slot
            store._by_object_key[defn.object_key] = slot

        return store

    async def read_by_name(self, volttron_name: str) -> float | None:
        """Read the active value of a point by its Volttron name.

        Raises:
            KeyError: If the point name is not found.
        """
        async with self._lock:
            slot = self._by_name[volttron_name]
            return self._active_value(slot)

    async def read_by_object_key(
        self, bacnet_object_type: str, index: int
    ) -> float | None:
        """Read the active value of a point by its BACnet object type and index.

        Raises:
            KeyError: If the (object_type, index) pair is not found.
        """
        async with self._lock:
            slot = self._by_object_key[(bacnet_object_type, index)]
            return self._active_value(slot)

    async def write_by_name(
        self,
        volttron_name: str,
        value: float,
        priority: int = 16,
    ) -> None:
        """Write a value to a writable point by its Volttron name.

        Raises:
            KeyError: If the point name is not found.
            TypeError: If the point is not writable.
        """
        async with self._lock:
            slot = self._by_name[volttron_name]
            self._write_slot(slot, value, priority)

    async def write_by_object_key(
        self,
        bacnet_object_type: str,
        index: int,
        value: float,
        priority: int = 16,
    ) -> None:
        """Write a value to a writable point by its BACnet object key.

        Raises:
            KeyError: If the (object_type, index) pair is not found.
            TypeError: If the point is not writable.
        """
        async with self._lock:
            slot = self._by_object_key[(bacnet_object_type, index)]
            self._write_slot(slot, value, priority)

    async def set_internal(self, volttron_name: str, value: float) -> None:
        """Set a point value directly, bypassing the priority array.

        Used by the simulation engine to update read-only sensor values
        (e.g., zone temperature, outdoor temperature).

        Raises:
            KeyError: If the point name is not found.
        """
        async with self._lock:
            slot = self._by_name[volttron_name]
            if slot.priority_array is not None:
                slot.priority_array = PriorityArray(default_value=value)
            slot.value = value

    async def read_all(self) -> dict[str, float | None]:
        """Return a snapshot of all point names and their active values."""
        async with self._lock:
            return {
                name: self._active_value(slot)
                for name, slot in self._by_name.items()
            }

    @property
    def point_names(self) -> list[str]:
        """Return all registered Volttron point names."""
        return list(self._by_name.keys())

    @property
    def definitions(self) -> list[PointDefinition]:
        """Return all registered point definitions."""
        return [slot.definition for slot in self._by_name.values()]

    @staticmethod
    def _active_value(slot: _PointSlot) -> float | None:
        if slot.priority_array is not None:
            return slot.priority_array.active_value()
        return slot.value

    @staticmethod
    def _write_slot(slot: _PointSlot, value: float, priority: int) -> None:
        if slot.priority_array is None:
            raise TypeError(
                f"Point '{slot.definition.volttron_name}' is not writable"
            )
        slot.priority_array.write(priority, value)

File: internal/points/store.go
package points

import (
	"fmt"
	"sync"
)

// PointStore is a thread-safe store for BACnet point values.
type PointStore struct {
	mu          sync.RWMutex
	values      map[string]*float64       // by volttron name
	objectIndex map[ObjectKey]string       // ObjectKey -> volttron name
	priorities  map[string]*PriorityArray  // writable points only
	definitions map[string]PointDefinition // by volttron name
}

// NewPointStore creates a store populated with point definitions.
// Default values are loaded into the value map and priority arrays.
func NewPointStore(defs []PointDefinition) *PointStore {
	ps := &PointStore{
		values:      make(map[string]*float64, len(defs)),
		objectIndex: make(map[ObjectKey]string, len(defs)),
		priorities:  make(map[string]*PriorityArray),
		definitions: make(map[string]PointDefinition, len(defs)),
	}

	for _, d := range defs {
		ps.definitions[d.VolttronName] = d
		ps.objectIndex[ObjectKey{ObjectType: d.BACnetObjectType, Index: d.Index}] = d.VolttronName

		if d.DefaultValue != nil {
			v := *d.DefaultValue
			ps.values[d.VolttronName] = &v
		}

		if d.Writable {
			pa := NewPriorityArray(d.DefaultValue)
			if d.DefaultValue != nil && d.WritePriority != nil {
				_ = pa.Write(*d.WritePriority, *d.DefaultValue)
			}
			ps.priorities[d.VolttronName] = pa
		}
	}

	return ps
}

// Read returns the current value of a point by its Volttron name.
func (ps *PointStore) Read(name string) (*float64, error) {
	ps.mu.RLock()
	defer ps.mu.RUnlock()

	if _, ok := ps.definitions[name]; !ok {
		return nil, fmt.Errorf("point %q not found", name)
	}
	v := ps.values[name]
	if v == nil {
		return nil, nil
	}
	cp := *v
	return &cp, nil
}

// ReadByKey returns the current value by BACnet object type and index.
func (ps *PointStore) ReadByKey(objectType string, index int) (*float64, error) {
	ps.mu.RLock()
	defer ps.mu.RUnlock()

	name, ok := ps.objectIndex[ObjectKey{ObjectType: objectType, Index: index}]
	if !ok {
		return nil, fmt.Errorf("object %s[%d] not found", objectType, index)
	}
	v := ps.values[name]
	if v == nil {
		return nil, nil
	}
	cp := *v
	return &cp, nil
}

// Write sets a point value through the priority array (external writes).
func (ps *PointStore) Write(name string, value float64, priority int) error {
	ps.mu.Lock()
	defer ps.mu.Unlock()

	def, ok := ps.definitions[name]
	if !ok {
		return fmt.Errorf("point %q not found", name)
	}
	if !def.Writable {
		return fmt.Errorf("point %q is not writable", name)
	}

	pa, ok := ps.priorities[name]
	if !ok {
		return fmt.Errorf("no priority array for point %q", name)
	}

	if err := pa.Write(priority, value); err != nil {
		return fmt.Errorf("write priority for %q: %w", name, err)
	}

	active := pa.ActiveValue()
	ps.values[name] = active
	return nil
}

// WriteByKey sets a point value by BACnet object type and index.
func (ps *PointStore) WriteByKey(objectType string, index int, value float64, priority int) error {
	ps.mu.Lock()
	defer ps.mu.Unlock()

	key := ObjectKey{ObjectType: objectType, Index: index}
	name, ok := ps.objectIndex[key]
	if !ok {
		return fmt.Errorf("object %s[%d] not found", objectType, index)
	}

	def := ps.definitions[name]
	if !def.Writable {
		return fmt.Errorf("point %q is not writable", name)
	}

	pa, ok := ps.priorities[name]
	if !ok {
		return fmt.Errorf("no priority array for point %q", name)
	}

	if err := pa.Write(priority, value); err != nil {
		return fmt.Errorf("write priority for %q: %w", name, err)
	}

	active := pa.ActiveValue()
	ps.values[name] = active
	return nil
}

// SetInternal sets a point value directly, bypassing the priority array.
// Used by the simulation engine to update sensor readings.
func (ps *PointStore) SetInternal(name string, value float64) error {
	ps.mu.Lock()
	defer ps.mu.Unlock()

	if _, ok := ps.definitions[name]; !ok {
		return fmt.Errorf("point %q not found", name)
	}
	v := value
	ps.values[name] = &v
	return nil
}

// ReadAll returns a snapshot of all current point values.
func (ps *PointStore) ReadAll() map[string]*float64 {
	ps.mu.RLock()
	defer ps.mu.RUnlock()

	out := make(map[string]*float64, len(ps.values))
	for k, v := range ps.values {
		if v != nil {
			cp := *v
			out[k] = &cp
		}
	}
	return out
}

// Definitions returns a copy of all point definitions.
func (ps *PointStore) Definitions() []PointDefinition {
	ps.mu.RLock()
	defer ps.mu.RUnlock()

	defs := make([]PointDefinition, 0, len(ps.definitions))
	for _, d := range ps.definitions {
		defs = append(defs, d)
	}
	return defs
}

// ReadFloat is a convenience method returning a float64 (0 if nil/missing).
func (ps *PointStore) ReadFloat(name string) float64 {
	v, err := ps.Read(name)
	if err != nil || v == nil {
		return 0
	}
	return *v
}

// ReadBool is a convenience method returning true if the value is > 0.
func (ps *PointStore) ReadBool(name string) bool {
	return ps.ReadFloat(name) > 0
}
